src/js/validator/validators/property.js

```diff
@@ -0,0 +1,131 @@
+'use strict';
+
+var validationError = require('../validation-error');
+var validateArity = require('./validation-utils').validateArity;
+var isRequired = require('./validation-utils').isRequired;
+
+function isOutOfRange(value, range) {
+  return ( range.minRange <= value ) && ( value <= range.maxRange );
+}
+
+function validateIntProp(prop, propSpec, nodeId) {
+  var errors = [];
+
+  function isInteger(n) {
+    return ( n === +n ) && ( n === (n|0) );
+  }
+
+  if (!isInteger(prop.val)) {
+    errors.push(validationError(nodeId, [propSpec.name, 'must be an integer'].join(' ')));
+  }
+
+  if (isOutOfRange(prop.value, propSpec.restrictions)) {
+    errors.push(validationError(nodeId, [
+      propSpec.name,
+      'must be within',
+      propSpec.restrictions.minRange,
+      'and',
+      propSpec.restrictions.maxRange
+    ].join(' ')));
+  }
+
+  return errors;
+}
+
+function validateFloatProp(prop, propSpec, nodeId) {
+  var errors = [];
+
+  if (typeof prop.val !== 'number') {
+    errors.push(validationError(nodeId, [propSpec.name, 'must be a float'].join(' ')));
+  }
+
+  if (isOutOfRange(prop.value, propSpec.restrictions)) {
+    errors.push(validationError(nodeId, [
+      propSpec.name,
+      'must be within',
+      propSpec.restrictions.minRange,
+      'and',
+      propSpec.restrictions.maxRange
+    ].join(' ')));
+  }
+
+  return errors;
+}
+
+function validateStringRestrictions(prop, restrictions, nodeId) {
+  var errors = [];
+
+  if (restrictions.items.indexOf(prop.val) === -1) {
+    errors.push(validationError(nodeId, [prop.val, 'is not a valid value for property', prop.key].join(' ')));
+  }
+
+  return errors;
+}
+
+function validateStringProp(prop, propSpec, nodeId) {
+  var errors = [];
+
+  function valueIsEnumerated(restrictions) {
+    return !!restrictions.items;
+  }
+
+  if (typeof prop.val !== 'string') {
+    errors.push(validationError(nodeId, [propSpec.name, 'must be a string'].join(' ')));
+  }
+
+  if (valueIsEnumerated(propSpec.restrictions)) {
+    errors.push.apply(errors, validateStringRestrictions(prop, propSpec.restrictions, nodeId));
+  }
+
+  return errors;
+}
+
+function validateBoolProp(prop, propSpec, nodeId) {
+  var errors = [];
+
+  if (typeof prop.value !== 'boolean') {
+    errors.push(validationError(nodeId, [propSpec.name, 'must be a boolean'].join(' ')));
+  }
+
+  return errors;
+}
+
+module.exports = function validateProperties(node, nodeSpec) {
+  var errors = [];
+  var nodeId = node.getId();
+  var nodePropMap = node.getProps();
+  var propSpecMap = nodeSpec.propertySpecMap;
+
+  Object.keys(propSpecMap).forEach(function(propName) {
+    var propSpec = propSpecMap[propName];
+    var props = nodePropMap[propSpec.name];
+
+    if (props) {
+      errors.push.apply(errors, validateArity(props, propSpec, nodeId));
```

Comment:
Yea I see what you mean now. I agree that we always want to return an array. It's not immediately obvious what the behavior of that line is and therefore the source of confusion. It's one of those querks of js. Looking online it looks like that line is another way of concatenating an array together so you'll need to ensure that errors don't get duplicated. 

https://davidwalsh.name/merge-arrays-javascript

As for readability I wonder if the ES6 spread operator would accomplish the same task? (On my phone so I can't test)

    errors.push(...[validate()])